xkcd.go

```diff
@@ -0,0 +1,83 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const xkcdURL = "https://xkcd.com/%d/info.0.json"
+
+func GetComic(comicNum int) (Comic, error) {
+	resp, err := http.Get(fmt.Sprintf(xkcdURL, comicNum))
+	if err != nil {
+		return Comic{}, fmt.Errorf("failed to GET comic %d: %s", comicNum, err)
+	}
+	defer resp.Body.Close()
+
+	var comicRetrieved Comic
+	if err := json.NewDecoder(resp.Body).Decode(&comicRetrieved); err != nil {
+		return Comic{}, fmt.Errorf("failed to GET comic %d: %s", comicNum, err)
+	}
+
+	return comicRetrieved, nil
+}
+
+func GetLatestComicNumber() (int, error) {
+	resp, err := http.Get("https://xkcd.com/info.0.json")
+	if err != nil {
+		return 0, fmt.Errorf("failed to perform GET request: %s", err)
+	}
+	defer resp.Body.Close()
+
+	var latest LatestComic
+	if err := json.NewDecoder(resp.Body).Decode(&latest); err != nil {
+		return 0, fmt.Errorf("failed to GET the latest comic number %d: %s", latest.Num, err)
+	}
+
+	fmt.Printf("Latest Comic Number is: %d", latest.Num)
+
+	return latest.Num, nil
+}
+
+func RetrieveComicsSequentially(start, end int) ([]Comic, time.Duration) {
+	var comics []Comic
+	startTime := time.Now()
+	for i := start; i <= end; i++ {
+		comic, err := GetComic(i)
+		if err != nil {
+			fmt.Printf("Error retrieving sequentially comic %d: %v\n", i, err)
+			continue
+		}
+		// fmt.Printf("comic number %d is %v\n", i, comic.Title)
+		comics = append(comics, comic)
+	}
+	duration := time.Since(startTime)
+	return comics, duration
+}
+
+func RetrieveComicsConcurrently(start, end int) ([]Comic, time.Duration) {
+	var comics []Comic
+	var wg sync.WaitGroup
+	startTime := time.Now()
+
+	for i := start; i <= end; i++ {
+		wg.Add(1)
+		go func(num int) {
+			defer wg.Done()
+			comic, err := GetComic(num)
+			if err != nil {
+				fmt.Printf("Error retrieving concurrently comic %d: %v\n", num, err)
+				return
+			}
+			// fmt.Printf("comic number %d is %v\n", i, comic.Title)
+			comics = append(comics, comic)
```

Comment:
It is unsafe to append to the same slice from different go routines as it leads to a race condition. It can sometimes work and other times the program can panic and shutdown.

Generally, it is unsafe to write to anything from different go routines without any form of synchronization, with some exceptions. It also boils down a lot to how your data and business logic is structured.

In this case, there are 2 possible issues:
- 2 go routines may try to add a new comic at the same index, in which case you may end up overriding one of them
- the slice doesn't have a size assigned on creation, so it will automatically resize based on some heuristics as you append to it. While resizing, you lose data or get panics.

There are multiple ways to deal with this issue. Here are a couple of them for your specific use case, from least to most suggested:
- use a [mutex](https://go.dev/tour/concurrency/9) to guard around the write operations. You would ```Lock/Unlock``` around the append. This would ensure you only write from one place at a time. Whilst this is the simplest solution, it would vastly impact your performance as only one of the go routines could write at a single time and all the others would have to wait. We generally try to steer clear of mutexes in highly concurrent scenarios because of this.
- you can take advantage of the data you have. You know from the beginning exactly how many comics you'll get, so you can just create the slice with enough capacity and then write at the corresponding index in each go routine. This is fine as you'll always be writing to each index from each go routine, even though they are all in the same slice. To achieve this, you would change line 62 to be ```comics := make([]Comic, 0, end + 1)``` and then instead of using append, you' do ```comics[i]=comic```. This is the best approach for your data, but I suggested it in the middle because it's data specific, not the most common approach you'd do and not applicable in more general cases.
- The most general thing you can do is to use a channel and have a single writer go routine. All the other routines would would get the comics and then just send them on the channel to the writer routine. As the writes are far quicker than the get requests, you'd still get maximum performance. This would look something like this:
```go
func RetrieveComicsConcurrently(start, end int) ([]Comic, time.Duration) {
	var comics []Comic
	comicCh := make(chan Comic, end)
	var getComicsWG sync.WaitGroup
	var writeWG sync.WaitGroup
	startTime := time.Now()

	// writer go routine
	writeWG.Add(1)
	go func() {
		defer writeWG.Done()
		for comic := range comicCh {
			comics = append(comics, comic)
		}
	}()

	for i := start; i <= end; i++ {
		getComicsWG.Add(1)
		go func(num int) {
			defer getComicsWG.Done()
			comic, err := GetComic(num)
			if err != nil {
				fmt.Printf("Error retrieving concurrently comic %d: %v\n", num, err)
				return
			}
			// fmt.Printf("comic number %d is %v\n", i, comic.Title)
			comicCh <- comic
		}(i)
	}

	// first wait to finish retrieving all comics
	getComicsWG.Wait()
	// then close the channel to signal to the writer go routine
	close(comicCh)
	// wait for the writer go routine to finish
	writeWG.Wait()

	duration := time.Since(startTime)
	return comics, duration
}
```